File: 78-subsets/78-subsets.cpp
class Solution {
public:
    
    set<vector<int>> st;
    void bt(vector<int> &temp, int i, vector<int>& nums)
    {
        st.insert(temp);
        if(i>=nums.size())return;
        
        temp.push_back(nums[i]);
        bt(temp,i+1,nums);
        temp.pop_back();
        
        bt(temp,i+1,nums);
    }
    
    vector<vector<int>> subsets(vector<int>& nums)
    {
        vector<vector<int>> ans;
        vector<int> temp;
        
        for(int i=0; i<nums.size(); i++)
        {
            temp.clear();
            bt(temp,i,nums);
        }
        
        for(auto it : st)
            ans.push_back(it);
        
        return ans;   
    }
};

File: 78-subsets/78. Subsets.java
class Solution {

    Set<List<Integer>> st = new HashSet<>();
    List<Integer> prev = new ArrayList<>();

    public void recur(int i, int[] nums) {
        if (i >= nums.length)
            return;

        // IO.println(i + " prev " + prev);

        // recur(i + 1, nums, prev);
        // recur(i + 1, nums, curr);

        for (int k = i; k < nums.length; k++) {
            //pick
            prev.add(nums[k]);
            st.add(new ArrayList<>(prev));
            recur(k+1, nums);

            // not pick
            prev.remove(prev.size()-1);
            recur(k+1, nums);
        }
    }

    public List<List<Integer>> subsets(int[] nums) {
        st.add(new ArrayList<>());
        recur(0, nums);

        IO.println(st);
        List<List<Integer>> ans = new ArrayList<>(st);
        return ans;
    }
}